fintrack/src/components/TransactionForm.js

```diff
@@ -0,0 +1,208 @@
+import React, { useState, useContext, useEffect } from 'react'
+import { useNavigate, useLocation } from 'react-router-dom'
+
+import { createTransaction, updateTransaction, deleteTransaction } from '../utils/api-transactions'
+
+import UserContext from '../context/UserContext'
+import { AuthContext } from '../context/AuthContext'
+
+function TransactionForm() {
+    const navigate = useNavigate()
+    const location = useLocation()
+
+    const { authToken } = useContext(AuthContext)
+    // get the user's categories and transaction types from context
+    const { authUser } = useContext(UserContext)
+
+    const [isEditMode, setIsEditMode] = useState(false)
+    const [account, setAccount] = useState(location.state?.selectedAccount)
+    const [transaction, setTransaction] = useState(location.state.transaction)
+    const [formData, setFormData] = useState({
+        account: '',
+        transactionType: '',
+        category: '',
+        amount: '',
+        description: ''
+    })
+    const [feedbackMessage, setFeedbackMessage] = useState('')
+
+    useEffect(() => {
+        console.log('use effect')
+        
+        const extractData = () => {
+            if (location.state?.transaction) {
+                setTransaction(location.state.transaction)
+                setIsEditMode(true)
+                setFormData({
+                    account: transaction.account._id,
+                    transactionType: transaction.transactionType._id,
+                    category: transaction.category._id,
+                    amount: transaction.amount,
+                    description: transaction.description
+                })
+            } else if (location.state?.selectedAccount) {
+                setAccount(location.state.selectedAccount)
+                setIsEditMode(false)
+                // account change functionality is not yet implemented
+                setFormData({
+                    account: account._id,
+                    transactionType: '',
+                    category: '',
+                    amount: '',
+                    description: ''
+                })
+            }
+        }
+        extractData()
+
+    }, [location, account, transaction])
```

Comment:
I am 100% confident this call to `useEffect` is not necessary because it does not synchronize our react app with an outside system. If account comes from `location.account`, we can use that directly. If form data comes from `account.id`, we can use that.

If `account.id` updates _just once_ after some parent-level API call, it's quite flaky to rely on that behavior and wipe the component state if it changes! Instead, we should _not_ render this component _at all_ until all the data it needs is ready.